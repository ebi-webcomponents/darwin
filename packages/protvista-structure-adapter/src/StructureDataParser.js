import ldFilter from "lodash-es/filter";
import ldMap from "lodash-es/map";
import ldForEach from "lodash-es/forEach";
import ldRemove from "lodash-es/remove";
import ParserHelper from "./ParserHelper";

const featureType = "PDBE_COVER";
const featureCategory = "STRUCTURE_COVERAGE";

export default class StructureDataParser {
  constructor() {
    this._pdbFeatures = [];
  }

  parseEntry(data) {
    this._parseValidEntry(data);
    return this._pdbFeatures;
  }

  get pdbFeatures() {
    return this._pdbFeatures;
  }

  static _getAllFeatureStructures(data) {
    let allFeatureStructures = [];
    const structures = ldFilter(data.dbReferences, reference => {
      return reference.type === "PDB";
    });
    allFeatureStructures = ldMap(structures, structure => {
      const beginEnd = structure.properties.chains
        ? ParserHelper.getBeginEnd(structure.properties.chains)
        : { start: 0, end: 0 };
      return {
        type: featureType,
        category: featureCategory,
        structures: [
          {
            description: ParserHelper.getDescription(structure.properties),
            start: beginEnd.begin,
            end: beginEnd.end,
            source: {
              id: structure.id,
              url: `http://www.ebi.ac.uk/pdbe-srv/view/entry/${structure.id}`
            }
          }
        ],
        start: beginEnd.begin,
        end: beginEnd.end
      };
    });
    return allFeatureStructures;
  }

  _getOverlapping(ftStructure, startEnd) {
    const overlapping = ldRemove(this._pdbFeatures, ftCoverage => {
      if (
        (ftCoverage.start <= ftStructure.start &&
          ftStructure.start <= ftCoverage.end) ||
        (ftCoverage.start <= ftStructure.end &&
          ftStructure.end <= ftCoverage.end) ||
        (ftStructure.start <= ftCoverage.end &&
          ftCoverage.end <= ftStructure.end)
      ) {
        /* eslint-disable no-param-reassign */
        startEnd.minStart = Math.min(
          startEnd.minStart,
          ftStructure.start,
          ftCoverage.start
        );
        /* eslint-disable no-param-reassign */
        startEnd.maxEnd = Math.max(
          startEnd.maxEnd,
          ftStructure.end,
          ftCoverage.end
        );
        return true;
      }
      return false;
    });
    return overlapping;
  }

  _parseValidEntry(data) {
    this._pdbFeatures = [];
    const allFeatureStructures = StructureDataParser._getAllFeatureStructures(
      data
    );

    ldForEach(allFeatureStructures, ftStructure => {
      const startEnd = { minStart: Number.MAX_SAFE_INTEGER, maxEnd: 1 };
      const overlapping = this._getOverlapping(ftStructure, startEnd);
      if (overlapping.length === 0) {
        this._pdbFeatures.push(ftStructure);
      } else {
        ftStructure.start = startEnd.minStart;
        ftStructure.end = startEnd.maxEnd;
        ldForEach(overlapping, overlap => {
          ftStructure.structures = ftStructure.structures.concat(
            overlap.structures
          );
        });
        this._pdbFeatures.push(ftStructure);
      }
    });
  }

  static _getStructuresHTML(structureList) {
    return `<ul>
            ${structureList
              .map(
                structure => `<li style="margin: 0.25rem 0"><a style="color:#FFF" href='${structure.source.url}' target='_blank'>
            ${structure.source.id}
        </a> (${structure.start}-${structure.end})</li>`
              )
              .join("")}
        </ul>`;
  }

  static formatTooltip(feature) {
    const structuresHTML = StructureDataParser._getStructuresHTML(
      feature.structures
    );
    return `${structuresHTML ? `<h5>Structures</h5>${structuresHTML}` : ``}`;
  }
}
