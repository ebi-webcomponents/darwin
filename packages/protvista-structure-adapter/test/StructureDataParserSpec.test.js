import StructureDataParser from "../src/StructureDataParser";
import { getEntryTestData } from "./UniProtEntryData";

describe("StructureDataParser", () => {
  it("should pass", () => {
    expect(true).toEqual(true);
  });

  it("should create a parser", () => {
    const aParser = new StructureDataParser();
    expect(aParser).toBeInstanceOf(StructureDataParser);
  });

  it("should parse mocked data", () => {
    const aParser = new StructureDataParser("P01234");
    const entryData = getEntryTestData();
    aParser.parseEntry(entryData);
    expect(JSON.stringify(aParser.pdbFeatures)).toMatchSnapshot();
  });
});
