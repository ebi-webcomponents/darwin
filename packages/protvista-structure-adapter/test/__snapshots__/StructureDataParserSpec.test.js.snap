// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructureDataParser should parse mocked data 1`] = `"[{\\"type\\":\\"PDBE_COVER\\",\\"category\\":\\"STRUCTURE_COVERAGE\\",\\"structures\\":[{\\"description\\":\\"Method: X-ray. Resolution: 1.50 A. \\",\\"end\\":0,\\"source\\":{\\"id\\":\\"1AAP\\",\\"url\\":\\"http://www.ebi.ac.uk/pdbe-srv/view/entry/1AAP\\"}}],\\"end\\":0},{\\"type\\":\\"PDBE_COVER\\",\\"category\\":\\"STRUCTURE_COVERAGE\\",\\"structures\\":[{\\"description\\":\\"Method: NMR. Chains: A=672-711. \\",\\"start\\":672,\\"end\\":711,\\"source\\":{\\"id\\":\\"1BA4\\",\\"url\\":\\"http://www.ebi.ac.uk/pdbe-srv/view/entry/1BA4\\"}},{\\"description\\":\\"Method: NMR. Chains: A=672-711. \\",\\"start\\":672,\\"end\\":711,\\"source\\":{\\"id\\":\\"1AML\\",\\"url\\":\\"http://www.ebi.ac.uk/pdbe-srv/view/entry/1AML\\"}},{\\"description\\":\\"Method: NMR. Chains: A=672-699. \\",\\"start\\":672,\\"end\\":699,\\"source\\":{\\"id\\":\\"1AMC\\",\\"url\\":\\"http://www.ebi.ac.uk/pdbe-srv/view/entry/1AMC\\"}},{\\"description\\":\\"Method: NMR. Chains: A=672-699. \\",\\"start\\":672,\\"end\\":699,\\"source\\":{\\"id\\":\\"1AMB\\",\\"url\\":\\"http://www.ebi.ac.uk/pdbe-srv/view/entry/1AMB\\"}}],\\"start\\":672,\\"end\\":711}]"`;
